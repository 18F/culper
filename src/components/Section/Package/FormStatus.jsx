import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Svg } from 'components/Form'
import AnimateReviewIcon from './AnimateReviewIcon'
import AnimateCheckmarkIcon from './AnimateCheckmarkIcon'

const FormStatus = ({
  progressWidth, isValid, isTransitioning, onTransitionEnd,
}) => (
  <div className="progress-container">
    <div className="review-icon">
      {isTransitioning ? (
        <img src="/img/review-checking.svg" alt="" />
      ) : (
        <AnimateReviewIcon valid={isValid} />
      )}
    </div>
    <div className="progress-outline">
      <div className="progress-default">
        <div
          className={classnames('progress', { transition: isTransitioning })}
          style={{ width: `${progressWidth}%` }}
          onTransitionEnd={onTransitionEnd}
        />
      </div>
    </div>
    <div className="icon-container">
      <span className="icon">
        {isTransitioning ? (
          <Svg src="/img/checkmark.svg" />
        ) : (
          <AnimateCheckmarkIcon valid={isValid} />
        )}
      </span>
    </div>
  </div>
)

FormStatus.propTypes = {
  progressWidth: PropTypes.number,
  isValid: PropTypes.bool,
  isTransitioning: PropTypes.bool,
  onTransitionEnd: PropTypes.func,
}

FormStatus.defaultProps = {
  progressWidth: 100,
  isValid: false,
  isTransitioning: false,
  onTransitionEnd: () => {},
}

export default FormStatus
