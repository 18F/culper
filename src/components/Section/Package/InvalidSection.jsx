import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

/** For use with top-level form sections only */
const InvalidSection = ({ section }) => (
  <div className="field">
    <div className="table expand">
      <div className="usa-alert usa-alert-error" role="alert">
        <div className="usa-alert-body">
          <h5 className="usa-alert-heading">
            {section.label}
          </h5>
          <ul>
            {section.subsections.map((ss) => {
              if (!ss.isValid) {
                return (
                  <li key={`package-review-errors-${ss.key}`}>
                    {ss.label}
                  </li>
                )
              }
              return null
            })}
          </ul>

          <Link to={`/form/${section.path}/review`}>
            <button
              type="button"
              className="back usa-button-outline"
            >
              Back to section
            </button>
          </Link>
        </div>
      </div>
    </div>
  </div>
)

InvalidSection.propTypes = {
  section: PropTypes.object.isRequired,
}

export default InvalidSection
