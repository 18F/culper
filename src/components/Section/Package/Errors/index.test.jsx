import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router'

import { PackageErrors } from './index'

import FormStatus from '../FormStatus'
import InvalidSection from '../InvalidSection'

describe('The PackageErrors component', () => {
  describe('with default props', () => {
    const component = mount(<PackageErrors />)

    it('renders without errors', () => {
      expect(component.exists()).toBe(true)
    })

    it('renders the FormStatus component', () => {
      const formStatus = component.find(FormStatus)
      expect(formStatus.length).toBe(1)
      expect(formStatus.prop('isValid')).toBe(false)
      expect(formStatus.prop('isTransitioning')).toBe(false)
    })
  })

  describe('with invalid sections', () => {
    const testProps = {
      formSections: [
        { key: '0', subsections: [{ key: '0.0', isValid: true }] },
        { key: '1', subsections: [{ key: '1.0', isValid: true }, { key: '1.1', isValid: false }] },
        { key: '2', subsections: [{ key: '2.0', isValid: false }] },
        { key: '3', subsections: [{ key: '3.0', subsections: [{ key: '3.0.0', isValid: true }] }] },
        { key: '4', subsections: [{ key: '4.0', subsections: [{ key: '4.0.0', isValid: false }] }] },
      ],
    }

    const component = mount(
      <MemoryRouter>
        <PackageErrors {...testProps} />
      </MemoryRouter>
    )

    it('renders the InvalidSection component for each invalid section', () => {
      expect(component.find(InvalidSection).length).toBe(3)
    })
  })
})
