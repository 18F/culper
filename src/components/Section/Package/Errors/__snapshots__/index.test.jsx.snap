// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PackageErrors component with default props renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="submission-status invalid"
  >
    <ReactMarkdown
      containerTagName="div"
      parserOptions={Object {}}
      renderers={
        Object {
          "link": [Function],
        }
      }
      source="# Some required fields are incomplete
Please fix the errors listed below."
    />
    <FormStatus
      isTransitioning={false}
      isValid={false}
      onTransitionEnd={[Function]}
      progressWidth={100}
    />
    <div
      className="invalid-form"
    >
      <ReactMarkdown
        containerTagName="div"
        parserOptions={Object {}}
        renderers={
          Object {
            "link": [Function],
          }
        }
        source="### List of incomplete sections
Use the list below to complete all incomplete sections or sections with errors."
      />
    </div>
  </div>,
  "nodes": Array [
    <div
      className="submission-status invalid"
    >
      <ReactMarkdown
        containerTagName="div"
        parserOptions={Object {}}
        renderers={
          Object {
            "link": [Function],
          }
        }
        source="# Some required fields are incomplete
Please fix the errors listed below."
      />
      <FormStatus
        isTransitioning={false}
        isValid={false}
        onTransitionEnd={[Function]}
        progressWidth={100}
      />
      <div
        className="invalid-form"
      >
        <ReactMarkdown
          containerTagName="div"
          parserOptions={Object {}}
          renderers={
            Object {
              "link": [Function],
            }
          }
          source="### List of incomplete sections
Use the list below to complete all incomplete sections or sections with errors."
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <PackageErrors
        formSections={Array []}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "formSections": Array [],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="submission-status invalid"
        >
          <ReactMarkdown
            containerTagName="div"
            parserOptions={Object {}}
            renderers={
              Object {
                "link": [Function],
              }
            }
            source="# Some required fields are incomplete
Please fix the errors listed below."
          />
          <FormStatus
            isTransitioning={false}
            isValid={false}
            onTransitionEnd={[Function]}
            progressWidth={100}
          />
          <div
            className="invalid-form"
          >
            <ReactMarkdown
              containerTagName="div"
              parserOptions={Object {}}
              renderers={
                Object {
                  "link": [Function],
                }
              }
              source="### List of incomplete sections
Use the list below to complete all incomplete sections or sections with errors."
            />
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="submission-status invalid"
        >
          <ReactMarkdown
            containerTagName="div"
            parserOptions={Object {}}
            renderers={
              Object {
                "link": [Function],
              }
            }
            source="# Some required fields are incomplete
Please fix the errors listed below."
          />
          <FormStatus
            isTransitioning={false}
            isValid={false}
            onTransitionEnd={[Function]}
            progressWidth={100}
          />
          <div
            className="invalid-form"
          >
            <ReactMarkdown
              containerTagName="div"
              parserOptions={Object {}}
              renderers={
                Object {
                  "link": [Function],
                }
              }
              source="### List of incomplete sections
Use the list below to complete all incomplete sections or sections with errors."
            />
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PackageErrors
    formSections={Array []}
  />,
}
`;
